### PR TITLE
RDM-6323: Implement Functional Tests for 'Validate a set of fields as Case worker' operation

### PR DESCRIPTION
**Pull request by SivaKanukolluQA**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/RDM-6323

### Change description ###
Implement Functional Tests for 'Validate a set of fields as Case worker' operation

**Does this PR introduce a breaking change?** (check one with "x")
[ ] Yes
[x ] No

### CODE REVIEW

File: src/aat/resources/features/F-050/F-050.feature [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    And it is submitted to call the [validation of a set of fields as Case worker] operation of [CCD Data Store]
+    Then a positive response is received
+    And the response [has the 200 return code]
+    And the response has all the details as expected
+
+  @Ignore
+  #To be completed on https://tools.hmcts.net/jira/browse/RDM-6654
+  @S-297
+  Scenario: must not validate when CMC ExternalID is not unique / already exists
+    Given a user with a detailed profile in CCD
```
Comment (MSancaktutar) [inline src/aat/resources/features/F-050/F-050.feature]: This won't work.

File: src/aat/resources/features/F-050/F-050.feature [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    Then a negative response is received
+    And the response [has the 403 return code]
+    And the response has all the details as expected
+
+  @S-299
+  Scenario: must return 4xx when request does not provide valid authentication credentials
+    Given a user with [a detailed profile in CCD]
+    When a request is prepared with appropriate values
+    And the request [uses a Case ID that doesn’t exist in CCD]
+    And the request [uses the invalid authorization]
```
Comment (MSancaktutar) [inline src/aat/resources/features/F-050/F-050.feature]: better like: uses an invalid authorization

File: src/aat/resources/features/F-050/S-298.td.json [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+          "PostTown": null,
+          "County": null,
+          "PostCode": null,
+          "Country": null
+        }
+      },
+      "event": {
+        "id": "CREATE",
+        "summary": "",
+        "description": ""
```
Comment (MSancaktutar) [inline src/aat/resources/features/F-050/S-298.td.json]: the content of the request body can be simplified by data reuse. you shouldn't repeat this body content in all test data. You can define it in a file and reuse it from there.

File: src/aat/resources/features/F-050/F-050.feature (hunk 1 above)
Comment (SivaKanukolluQA) [reply src/aat/resources/features/F-050/F-050.feature]: Updated this to : Scenario: must not validate when CMC ExternalID is not unique or already exists.  This is an @Ignored scenario and will implement later.

File: src/aat/resources/features/F-050/F-050.feature (hunk 2 above)
Comment (SivaKanukolluQA) [reply src/aat/resources/features/F-050/F-050.feature]: Implemented:     And the request [uses an invalid authorization]

File: src/aat/resources/features/F-050/S-298.td.json (hunk 3 above)
Comment (SivaKanukolluQA) [reply src/aat/resources/features/F-050/S-298.td.json]: Implemented. Created a request file : F-050_Test_Data_Base.td.json

File: build.gradle [hunk 4]
```diff
@@ -478,9 +478,8 @@ jacoco {
 
 sonarqube {
     properties {
-        property "sonar.exclusions", "build/generated-sources/**/*.java,**/service/doclink/**"
-        property "sonar.coverage.exclusions", "src/main/java/uk/gov/hmcts/ccd/data/definition/DefaultCaseDefinitionRepository.java,src/main/java/uk/gov/hmcts/ccd/AppInsights.java," +
-                "**/DocLinksRestoreEndpoint.java,**/service/doclink/**"
+        property "sonar.exclusions", "build/generated-sources/**/*.java"
+        property "sonar.coverage.exclusions", "src/main/java/uk/gov/hmcts/ccd/data/definition/DefaultCaseDefinitionRepository.java,src/main/java/uk/gov/hmcts/ccd/AppInsights.java"
```
Comment (MSancaktutar) [inline build.gradle]: I was expecting no change in build.gradle for a FTA ticket

File: charts/ccd-data-store-api/Chart.yaml [hunk 5]
```diff
@@ -2,7 +2,7 @@ description: Helm chart for the HMCTS CCD Data Store
 name: ccd-data-store-api
 apiVersion: v1
 home: https://github.com/hmcts/ccd-data-store-api
-version: 1.2.7
+version: 1.2.8
```
Comment (MSancaktutar) [inline charts/ccd-data-store-api/Chart.yaml]: Changes in Chart.yml?

File: dependency-check-suppressions.xml [hunk 6]
```diff
@@ -5,6 +5,7 @@
         <notes>We do not use: Spring Framework 5.0.5.RELEASE + Spring Security (any version), see https://pivotal.io/security/cve-2018-1258</notes>
         <cve>CVE-2018-1258</cve>
     </suppress>
+
```
Comment (MSancaktutar) [inline dependency-check-suppressions.xml]: why change this?

File: src/aat/resources/features/F-050/F-050.feature [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+    When a request is prepared with appropriate values
+    And the request [uses a value that exceeds the field max limit]
+    And it is submitted to call the [validation of a set of fields as Case worker] operation of [CCD Data Store]
+    Then a negative response is received
+    And the response [has the 422 return code]
+    And the response has all other details as expected
+
+  @S-300
+  Scenario: must return 403 when request provides authentic credentials without authorised access to the operation
+    Given a user with [a detailed profile in CCD]
```
Comment (MSancaktutar) [inline src/aat/resources/features/F-050/F-050.feature]: "an active profile in CCD" would be better

File: src/aat/resources/features/F-050/F-050.feature [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+  #To be completed on https://tools.hmcts.net/jira/browse/RDM-6654
+  @S-297
+  Scenario: must not validate when CMC ExternalID is not unique or already exists
+    Given a user with [a detailed profile in CCD]
+    When a request is prepared with appropriate values
+    And the request [uses the existing CMC External Id]
+    And it is submitted to call the [validation of a set of fields as Case worker] operation of [CCD Data Store]
+    Then a negative response is received
+    And the response [has the 409 return code]
+    And the response has all other details as expected
```
Comment (MSancaktutar) [inline src/aat/resources/features/F-050/F-050.feature]: Make sure there is an EOL at the of the file

File: src/aat/resources/features/F-050/F-050.feature [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+    And it is submitted to call the [validation of a set of fields as Case worker] operation of [CCD Data Store]
+    Then a negative response is received
+    And the response [has the 403 return code]
+    And the response has all other details as expected
+
+  @S-299
+  Scenario: must return 4xx when request does not provide valid authentication credentials
+    Given a user with [a detailed profile in CCD]
+    When a request is prepared with appropriate values
+    And the request [uses a Case ID that doesn’t exist in CCD]
```
Comment (MSancaktutar) [inline src/aat/resources/features/F-050/F-050.feature]: irrelevant to scenario title

File: src/aat/resources/features/F-050/F-050.feature [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+    And the response [has the 403 return code]
+    And the response has all other details as expected
+
+  @Ignore
+  #To be completed on https://tools.hmcts.net/jira/browse/RDM-6654
+  @S-297
+  Scenario: must not validate when CMC ExternalID is not unique or already exists
+    Given a user with [a detailed profile in CCD]
+    When a request is prepared with appropriate values
+    And the request [uses the existing CMC External Id]
```
Comment (MSancaktutar) [inline src/aat/resources/features/F-050/F-050.feature]: "uses an existing CMC External Id"

File: src/aat/resources/features/F-050/S-300.td.json [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+  },
+  "expectedResponse": {
+    "_extends_": "Common_403_Response",
+    "headers": {
+      "Transfer-Encoding" : "chunked"
+    },
+    "body": {
+      "timestamp" : "[[ANY]]",
+      "message" : "Access Denied",
+      "path" : "[[ANY]]"
```
Comment (MSancaktutar) [inline src/aat/resources/features/F-050/S-300.td.json]: please remove the common fields that are now coming from base object.

File: src/aat/resources/features/F-050/S-299.td.json [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+  },
+  "expectedResponse": {
+    "_extends_": "Common_403_Response",
+    "headers": {
+      "Transfer-Encoding" : "chunked"
+    },
+    "body": {
+      "timestamp" : "[[ANY]]",
+      "message" : "Access Denied",
+      "path" : "[[ANY]]"
```
Comment (MSancaktutar) [inline src/aat/resources/features/F-050/S-299.td.json]: please check and remove tall he common fields that are now coming from base object.

File: src/aat/resources/features/F-050/S-298.td.json [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+      "details" : {
+        "field_errors" : [
+          {
+            "id": "PhoneUKField",
+            "message": "The data entered is not valid for this type of field, please delete and re-enter using only valid data"
+          }
+        ]
+      },
+      "callbackErrors" : null,
+      "callbackWarnings" : null
```
Comment (MSancaktutar) [inline src/aat/resources/features/F-050/S-298.td.json]: please check and remove tall he common fields that are now coming from base object.

File: src/aat/resources/features/F-050/S-298.td.json [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+    }
+  },
+  "expectedResponse": {
+    "_extends_": "Common_422_Response",
+    "headers": {
+      "Transfer-Encoding" : "chunked"
+    },
+    "body": {
+      "exception" : "uk.gov.hmcts.ccd.endpoint.exceptions.CaseValidationException",
+      "timestamp" : "[[ANY]]",
```
Comment (MSancaktutar) [inline src/aat/resources/features/F-050/S-298.td.json]: please check and remove tall he common fields that are now coming from base object.

File: src/aat/resources/features/F-050/S-298.td.json [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+      "message" : "Case data validation failed",
+      "path" : "[[ANY]]",
```
Comment (MSancaktutar) [inline src/aat/resources/features/F-050/S-298.td.json]: please check and remove tall he common fields that are now coming from base object.

File: src/aat/resources/features/F-050/S-297.td.json [hunk 16]
```diff
@@ -0,0 +1,21 @@
...
+  ],
+
+  "expectedResponse": {
+    "_extends_": "Common_409_Response",
+    "body": {
+      "exception" : "uk.gov.hmcts.ccd.endpoint.exceptions.CaseValidationException",
+      "timestamp" : "[[ANY]]",
+      "path" : "[[ANY]]",
+      "callbackErrors" : null,
+      "callbackWarnings" : null
```
Comment (MSancaktutar) [inline src/aat/resources/features/F-050/S-297.td.json]: please check and remove tall he common fields that are now coming from base object.

File: src/aat/resources/features/F-050/F-050.feature (hunk 7 above)
Comment (SivaKanukolluQA) [reply src/aat/resources/features/F-050/F-050.feature]: Updated:  Given a user with [an active profile in CCD]

File: src/aat/resources/features/F-050/F-050.feature (hunk 9 above)
Comment (SivaKanukolluQA) [reply src/aat/resources/features/F-050/F-050.feature]: Actually this scenario for 401 status code. As of now, we are not able to reproduce this.
So updated the tile like this :    Scenario: must return negative response when request does not provide valid authentication credentials.

File: src/aat/resources/features/F-050/F-050.feature (hunk 10 above)
Comment (SivaKanukolluQA) [reply src/aat/resources/features/F-050/F-050.feature]: Updated:    And the request [uses an existing CMC External Id]

File: src/aat/resources/features/F-050/S-300.td.json (hunk 11 above)
Comment (SivaKanukolluQA) [reply src/aat/resources/features/F-050/S-300.td.json]: Removed Timestamp & path.

File: src/aat/resources/features/F-050/S-299.td.json (hunk 12 above)
Comment (SivaKanukolluQA) [reply src/aat/resources/features/F-050/S-299.td.json]: Removed timestamp & path.

File: src/aat/resources/features/F-050/S-298.td.json (hunk 13 above)
Comment (SivaKanukolluQA) [reply src/aat/resources/features/F-050/S-298.td.json]: Removed callback error & warning.

File: src/aat/resources/features/F-050/S-298.td.json (hunk 15 above)
Comment (SivaKanukolluQA) [reply src/aat/resources/features/F-050/S-298.td.json]: Removed time stamp and path.

**Comment by SivaKanukolluQA** [comment]
Not required any more. Created new PR from RDM-6329.

**Pull request closed without merging**